zerver/tests/test_typing.py

```diff
@@ -0,0 +1,156 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+from typing import Any, Dict, List
+from six import string_types
+
+from zerver.models import get_user_profile_by_email
+
+from zerver.lib.test_helpers import ZulipTestCase, tornado_redirected_to_list, get_display_recipient
+
+class TypingNotificationTest(ZulipTestCase):
+    def test_send_notification_to_self_event(self):
+        # type: () -> None
+        """
+        Sending typing notification to yourself
+        is successful.
+        """
+        email = "hamlet@zulip.com"
+
+        events = [] # type: List[Dict[str, Any]]
+        with tornado_redirected_to_list(events):
+            result = self.client_post("/api/v1/typing", {"to": email},
+                                        **self.api_auth(email))
+        self.assert_json_success(result)
+        self.assertTrue(len(events) == 1)
+
+        event = events[0]['event']
+        display_recipient = get_display_recipient(event['recipient'])
+        if not isinstance(display_recipient, string_types):
+            event_recipient_emails = set(user['email'] for user in display_recipient)
+        else:
+            event_recipient_emails = set([display_recipient])
+
+        self.assertTrue(event['sender'] == get_user_profile_by_email(email))
+        self.assertTrue(event_recipient_emails == set([email]))
+        self.assertTrue(event['type'] == 'typing_notification')
+
+    def test_send_notification_to_another_user_event(self):
+        # type: () -> None
+        """
+        Sending typing notification to another user
+        is successful.
+        """
+        sender = "hamlet@zulip.com"
+        recipient = "othello@zulip.com"
+        events = [] # type: List[Dict[str, Any]]
+        with tornado_redirected_to_list(events):
+            result = self.client_post("/api/v1/typing", {"to": recipient},
+                                        **self.api_auth(sender))
+        self.assert_json_success(result)
+        self.assertTrue(len(events) == 1)
+
+        event = events[0]['event']
+        display_recipient = get_display_recipient(event['recipient'])
+        if not isinstance(display_recipient, string_types):
+            event_recipient_emails = set(user['email'] for user in display_recipient)
+        else:
+            event_recipient_emails = set([display_recipient])
+
+        self.assertTrue(event['sender'] == get_user_profile_by_email(sender))
+        self.assertTrue(event_recipient_emails == set([recipient]))
+        self.assertTrue(event['type'] == 'typing_notification')
+
+    def test_send_notification_without_reciever(self):
+        # type: () -> None
+        """
+        Sending typing notification without receiver fails
+        """
+        sender = "hamlet@zulip.com"
+        result = self.client_post("/api/v1/typing", {},
+                                        **self.api_auth(sender))
```

Comment:
The indentation here is wrong; `**self.api_auth` should be aligned with the `"/api/...`; there's a bunch more cases of this below.